packages/material-ui/src/ButtonBase/ButtonBase.test.js

```diff
@@ -374,6 +374,52 @@ describe('<ButtonBase />', () => {
           button.querySelectorAll('.ripple-visible .child:not(.child-leaving)'),
         ).to.have.lengthOf(0);
       });
+
+      it('should not crash when changes enableRipple from false to true', () => {
+        function App() {
+          /** @type {React.MutableRefObject<import('./ButtonBase').ButtonBaseActions | null)> | null} */
+          const buttonRef = React.useRef(null);
+          const [enableRipple, setRipple] = React.useState(false);
+
+          React.useEffect(() => {
+            if (buttonRef.current) {
```

Comment:
This is the type of defensive checks I was working about that TypeScript encourage and that might hide root issues. @eps1lon what do you think?

Reply:
This one is not hiding issue of test, you can check it and without changes this test will still crash

Reply:
Agree, I was wondering about the pattern in general, what should be our "baseline" (default approach) for this concern :)

Reply:
Do we even need this check here? What happens if `App` mounts but the ref isn't instantiated?

Reply:
This check if for typescript compiler, to get rid of `@ts-ignore`. Actually it is impossible inside the `useEffect` (maybe only if somebody will specifically delete dom element after react commit phase and before effects are run?) 

Right here it is not needed, but I am sure that in the code we have to guard the refs, even if it is hide bugs (IMHO such guards can't hide bugs, because e.g. in this particular issue element will not gain ripple focus if check was done - it is a bug as well, but no so critical as crash)

Reply:
>  it is a bug as well, but no so critical as crash)

So we agree that it shouldn't be guarded against. There was a defect in the code which we wouldn't have been able to detect with a defensive check.

As long as typescript is only a type checker + transpiler I don't care about `@ts-ignore` in JSX. TypeScript is particularly bad with React. If it complains, we disable it. Just like with false positives in lint rules.

Reply:
We still need to get rid of this check in the test or handle the else case.

Reply:
👌 ok I will add throwing error in the else branch. 
Cannot agree — we would be able. If there will be a guard when enableRipple changes you simple not starting rippleEffect. It is an issue that can be noticed and fixed in the same way as current one. 

**But** if some user will notice crash in the button component — it will spoil the trust to material-ui as _qualified and reliable tool_. 

I have strong opinion that if something can crash — you must not let it crash. Even if you will use more reliable type system like ReasonML will complain about refs, because we cannot trust DOM, like we cannot trust users and we cannot trust developers who uses our components. 

I can propose you the following api 
```js
assertReactDomRef(ref, ref => ref.focus())
```

Which will assert ref is not null and if it’s missing warn user and ask to raise an issue. Thus our component will even look more clever :)

Reply:
As far as I understand the issue, this is an internal consideration. Regarding the impact userland, I think that a fast feedback loop should be preferred.

Reply:
@dmtrKovalenko I love the new throw error in the branch, it's a great edge in case the action prop stop working as expected, we get a clear error message in the test. Smart! In an older test, we were using `// @ts-ignore`, which sounds great too.

> But if some user will notice crash in the most basic Button component — it will spoil the trust to material-ui as qualified and reliable tool.

From my perspective, the ref effect should always run before a layout effect which also runs before an effect. The ref should always be defined. If it's not, then there is a deeper issue, that a defensive logic will hide, make it harder to uncover.
So in userland, I would recommend the usage of `// @ts-ignore`